projection/validator_delegation/view/validators.go

```diff
@@ -29,43 +30,288 @@ func NewValidatorsView(handle *rdb.Handle) Validators {
+
+		err := view.setValidatorRecordHeightUpperBound(row)
+		if err != nil {
+			return fmt.Errorf("error updating Validator.Height upper bound: %v", err)
+		}
+
+		err = view.Insert(row)
+		if err != nil {
+			return fmt.Errorf("error inserting a new record for this validator: %v", err)
+		}
+
+	}
+
 	return nil
 }
 
+func (view *ValidatorsView) checkIfValidatorRecordExistByHeightLowerBound(row ValidatorRow) (bool, error) {
+
+	sql, sqlArgs, err := view.rdb.StmtBuilder.
+		Select(
+			"COUNT(*)",
+		).
+		From("view_vd_validators").
+		Where(
+			"operator_address = ? AND height = ?",
+			row.OperatorAddress,
+			fmt.Sprintf("[%v,)", row.Height),
+		).
+		ToSql()
+	if err != nil {
+		return false, fmt.Errorf("error building sql to check validator at specific height: %v: %w", err, rdb.ErrPrepare)
+	}
+
+	var count int64
+	if err = view.rdb.QueryRow(sql, sqlArgs...).Scan(
+		&count,
+	); err != nil {
+		return false, fmt.Errorf("error scanning count: %v: %w", err, rdb.ErrQuery)
+	}
+
+	return count == 1, nil
+}
+
 func (view *ValidatorsView) Delete(row ValidatorRow) error {
 
+	return view.setValidatorRecordHeightUpperBound(row)
+}
+
+func (view *ValidatorsView) setValidatorRecordHeightUpperBound(row ValidatorRow) error {
+
+	// Set the upper bound for record height: `[<PREVIOUS-LOWER-BOUND>, row.Height)`.
+	sql, sqlErr := view.rdb.StmtBuilder.ReplacePlaceholders(`
+		UPDATE view_vd_validators
+		SET height = int8range(lower(height), ?, '[)')
+		WHERE operator_address = ? AND height @> ?::int8
+	`)
+	if sqlErr != nil {
+		return fmt.Errorf("error building validator upper(height) update sql: %v: %w", sqlErr, rdb.ErrBuildSQLStmt)
+	}
+	sqlArgs := []interface{}{
+		row.Height,
+		row.OperatorAddress,
+		row.Height,
+	}
+
+	if _, execErr := view.rdb.Exec(sql, sqlArgs...); execErr != nil {
+		return fmt.Errorf("error executing validator upper(height)update sql: %v: %w", execErr, rdb.ErrWrite)
+	}
+
 	return nil
 }
 
 func (view *ValidatorsView) FindByOperatorAddr(operatorAddress string, height int64) (ValidatorRow, bool, error) {
 
-	// TODO handle the error when validator is NOT FOUND
+	row, found, err := view.findBy(&operatorAddress, nil, height)
+	if err != nil {
+		return ValidatorRow{}, false, fmt.Errorf("error finding validator by OperatorAddr: %v", err)
+	}
 
-	return ValidatorRow{}, true, nil
+	return row, found, nil
 }
 
 func (view *ValidatorsView) FindByConsensusNodeAddr(consensusNodeAddress string, height int64) (ValidatorRow, bool, error) {
 
-	// TODO handle the error when validator is NOT FOUND
+	row, found, err := view.findBy(nil, &consensusNodeAddress, height)
+	if err != nil {
+		return ValidatorRow{}, false, fmt.Errorf("error finding validator by OperatorAddr: %v", err)
+	}
 
-	return ValidatorRow{}, true, nil
+	return row, found, nil
 }
 
-func (view *ValidatorsView) FindByTendermintAddr(tendermintAddress string, height int64) (ValidatorRow, error) {
+func (view *ValidatorsView) findBy(
```

Comment:
is it intended to make it private?

Reply:
Yes. 

The `FindByOperatorAddr` and `FindByConsensusNodeAddr` will invoke `findBy` internally.

`FindByTendermintAddr` is not used in implementation, so I simply remove it.